runtime/doc/lsp.txt

```diff
@@ -0,0 +1,359 @@
+				You can pass any name as server_name.
+				If you want to use multiple language servers
+				for one filetype, be sure to add server_name.
+
+			`cmd` (dictionary): Executing language server command.
+				`execute_path` (string):
+					Executing language server command path.
+				`args` (array):
+					Command arguments.
+
+			`server_config` (dictionary): Optional configuration for a server.
+
+	Example: ~
+>
+		" C, C++ example
+		call lsp#add_server_config({
+		      \ 'filetype': ['c', 'cpp'],
+		      \ 'server_name': 'cland',
+		      \ 'cmd': { 'execute_path': 'cland', 'args': ['-background-index'] }})
+<
+>
+		" Go example
+		call lsp#add_server_config({
+		      \ 'filetype': 'go',
+		      \ 'server_name': 'gopls',
+		      \ 'cmd': { 'execute_path': 'gopls', 'args': [] }})
+<
+>
+		" Python example
+		call lsp#add_server_config({
+		      \ 'filetype': 'python',
+		      \ 'server_name': 'pyls',
+		      \ 'cmd': { 'execute_path': 'pyls', 'args': [] }})
+<
+>
+		" Rust example
+		call lsp#add_server_config({
+		      \ 'filetype': 'rust',
+		      \ 'server_name': 'rls',
+		      \ 'cmd': { 'execute_path': 'rustup', 'args': ['run', 'stable', 'rls'] },
+		      \ 'server_config': {
+		      \   'clippy_preference': 'on',
+		      \   'all_targets': v:false,
+		      \   'build_on_save': v:true,
+		      \   'wait_to_build': 0
+		      \ }})
+<
+
+================================================================================
+                                                                  *lsp-request*
+Requests
+
+To make a request to the language server client running in Nvim, the following
+API is defined:
+>
+	:lua vim.lsp.request(method, arguments, callback, bufnr, filetype)
+	:lua vim.lsp.request_async(method, arguments, callback, bufnr, filetype)
+<
+
+                                                            *vim.lsp.request()*
+vim.lsp.request({method} [, {arguments} [, {callback} [, {bufnr} [, {filetype} [, {server_name}]]]]]):
+	Creates and sends a request to the server for [filetype]. Will fill in any
+	missing required arguments as defined by the language server protocol. Will
+	call the built-in callback for the request (if you set that and it exists), unless an
+	[callback] is specified, in which case that will be called upon completion
+	of the request.
+
+	Returns the request's response data.
+
+	Parameters: ~
+		{method}: The name of the request, as specified by the
+		language server protocol.
+			Example: 'textDocument/hover'
+
+		{arguments}: The arguments that should be passed to the language
+		server. If any required keys are missing, Nvim will fill those in
+		with reasonable default values.
+			Example (with 'textDocument/hover'):
+				Passed in:
+					{} or |v:null|
+
+				Sent:
+
+					{
+					    'textDocument': '/home/user/test/example.lua',
+					    'position': { 'line': 13, 'character': 17 }
+					}
+
+		{callback}: An optional callback that will be executed
```

Comment:
though this API seems over all a little weird. as `vim.lsp.request` returns the result, perhaps the default should be no callback called?

Reply:
No, `vim.lsp.request` call callback method same as `vim.lsp.request_async` now.
Should we make `vim.lsp.request` not call callback and remove callback from arguments ?

Reply:
I remove callback argument.
https://github.com/neovim/neovim/pull/10222/commits/181984031d653d93f4706b517cccb66923770258

Reply:
@bfredl I remove callback argument.
https://github.com/neovim/neovim/pull/10222/commits/181984031d653d93f4706b517cccb66923770258